File: eval-reviews/2026-09-12-chained/workflow-recovery/typescript/terra/r1/starter/workflow.ts
/** Deterministic durable, in-memory workflow runner. */
export const TIME_LIMIT = 2_147_483_647;
export class DomainError extends Error { readonly code: string; constructor(code: string) { super(code); this.code = code; } }
function require(condition: unknown, code = "INVALID_INPUT"): asserts condition { if (!condition) throw new DomainError(code); }
function fields(value: unknown, required: string[], optional: string[] = []): Record<string, unknown> {
  require(value !== null && typeof value === "object" && !Array.isArray(value));
  const object = value as Record<string, unknown>;
  require(required.every(k => Object.hasOwn(object, k)) && Object.keys(object).every(k => required.includes(k) || optional.includes(k)));
  return object;
}
function integer(value: unknown, low: number, high: number): number {
  require(typeof value === "number" && Number.isInteger(value) && value >= low && value <= high);
  return value;
}
function identifier(value: unknown): string { require(typeof value === "string" && /^[A-Za-z0-9_-]{1,64}$/.test(value)); return value; }

export interface StepDefinition { readonly id: string; readonly needs: readonly string[]; readonly amount: number; readonly failures: number; }
export type Command =
  | { op: "start" | "cancel"; run: string }
  | { op: "advance"; by: number }
  | { op: "tick"; crashAt: "after_begin" | "after_call" | null }
  | { op: "observe" | "crash" | "restart" };
export interface Workflow { readonly steps: readonly StepDefinition[]; readonly commands: readonly Command[]; readonly maxAttempts: number; readonly retryDelay: number; }
function parseStep(raw: unknown): StepDefinition {
  const o = fields(raw, ["id", "needs", "amount"], ["failures"]);
  require(Array.isArray(o.needs)); const needs = o.needs.map(identifier); require(new Set(needs).size === needs.length);
  return { id: identifier(o.id), needs, amount: integer(o.amount, 1, 1_000_000), failures: integer(Object.hasOwn(o, "failures") ? o.failures : 0, 0, 100) };
}
function parseCommand(raw: unknown): Command {
  require(raw !== null && typeof raw === "object" && !Array.isArray(raw)); const o = raw as Record<string, unknown>;
  if (o.op === "start" || o.op === "cancel") { fields(raw, ["op", "run"]); return { op: o.op, run: identifier(o.run) }; }
  if (o.op === "advance") { fields(raw, ["op", "by"]); return { op: "advance", by: integer(o.by, 0, TIME_LIMIT) }; }
  if (o.op === "tick") { fields(raw, ["op"], ["crash_at"]); const c = Object.hasOwn(o, "crash_at") ? o.crash_at : null; require(c === null || c === "after_begin" || c === "after_call"); return { op: "tick", crashAt: c }; }
  if (o.op === "observe" || o.op === "crash" || o.op === "restart") { fields(raw, ["op"]); return { op: o.op }; }
  throw new DomainError("INVALID_INPUT");
}
function validateGraph(steps: readonly StepDefinition[]): void {
  const ids = new Set(steps.map(s => s.id)); require(ids.size === steps.length, "DUPLICATE_STEP");
  require(steps.every(s => s.needs.every(d => ids.has(d))), "UNKNOWN_DEPENDENCY");
  const done = new Set<string>(); let todo = [...steps];
  while (todo.length) { const ready = todo.filter(s => s.needs.every(d => done.has(d))); require(ready.length, "DEPENDENCY_CYCLE"); ready.forEach(s => done.add(s.id)); todo = todo.filter(s => !done.has(s.id)); }
}
export function parseWorkflow(raw: unknown): Workflow {
  const o = fields(raw, ["steps", "commands"], ["max_attempts", "retry_delay"]); require(Array.isArray(o.steps) && o.steps.length && Array.isArray(o.commands));
  const w: Workflow = { steps: o.steps.map(parseStep), commands: o.commands.map(parseCommand), maxAttempts: integer(Object.hasOwn(o, "max_attempts") ? o.max_attempts : 3, 1, 10), retryDelay: integer(Object.hasOwn(o, "retry_delay") ? o.retry_delay : 2, 1, 1_000_000) };
  validateGraph(w.steps); return w;
}

type StepStatus = "pending" | "running" | "succeeded" | "failed" | "blocked" | "cancelled";
type RunStatus = "active" | "succeeded" | "failed" | "cancelling" | "cancelled";
export interface StepState { id: string; status: StepStatus; attempts: number; ready_at: number; }
export interface RunState { id: string; status: RunStatus; cancel_requested: boolean; steps: StepState[]; }
type ActionKey = readonly [string, string];
interface ServiceCall { kind: "execute" | "lookup"; key: ActionKey; attempt: number; outcome: "applied" | "replayed" | "transient" | "found" | "missing"; }
interface Effect { key: ActionKey; amount: number; }
type ExecuteOutcome = "applied" | "replayed" | "transient";

export class MockService {
  readonly calls: ServiceCall[] = []; readonly effects: Effect[] = [];
  private readonly effectsByRun = new Map<string, Map<string, Effect>>();
  private readonly executeCounts = new Map<string, Map<string, number>>();
  private effect(key: ActionKey): Effect | undefined { return this.effectsByRun.get(key[0])?.get(key[1]); }
  execute(key: ActionKey, amount: number, attempt: number, failures: number): ExecuteOutcome {
    let outcome: ExecuteOutcome;
    if (this.effect(key)) outcome = "replayed";
    else { const counts = this.executeCounts.get(key[0]) ?? new Map<string, number>(); this.executeCounts.set(key[0], counts); const n = (counts.get(key[1]) ?? 0) + 1; counts.set(key[1], n);
      if (n <= failures) outcome = "transient";
      else { outcome = "applied"; const effect: Effect = { key: [key[0], key[1]], amount }; const effects = this.effectsByRun.get(key[0]) ?? new Map<string, Effect>(); this.effectsByRun.set(key[0], effects); effects.set(key[1], effect); this.effects.push(effect); }
    }
    this.calls.push({ kind: "execute", key: [key[0], key[1]], attempt, outcome }); return outcome;
  }
  lookup(key: ActionKey, attempt: number): "found" | "missing" { const outcome = this.effect(key) ? "found" : "missing"; this.calls.push({ kind: "lookup", key: [key[0], key[1]], attempt, outcome }); return outcome; }
}
export interface Snapshot { now: number; up: boolean; runs: RunState[]; }
export interface SimulationResult { observations: Snapshot[]; final: Snapshot & { calls: ServiceCall[]; effects: Effect[] }; }
interface Action { run: RunState; step: StepState; definition: StepDefinition; }

export class Simulator {
  now = 0; up = true; readonly runs: RunState[] = []; readonly service = new MockService(); readonly observations: Snapshot[] = [];
  readonly workflow: Workflow;
  constructor(workflow: Workflow) { this.workflow = workflow; }
  private action(run: RunState, index: number): Action { return { run, step: run.steps[index]!, definition: this.workflow.steps[index]! }; }
  selectAction(): Action | undefined {
    for (const run of this.runs) { const i = run.steps.findIndex(s => s.status === "running"); if (i >= 0) return this.action(run, i); }
    for (const run of this.runs) if (run.status === "active") {
      const succeeded = new Set(run.steps.filter(s => s.status === "succeeded").map(s => s.id));
      for (let i = 0; i < run.steps.length; i++) { const step = run.steps[i]!, def = this.workflow.steps[i]!; if (step.status === "pending" && step.ready_at <= this.now && def.needs.every(d => succeeded.has(d))) return this.action(run, i); }
    }
    return undefined;
  }
  private checkpoint(crashAt: "after_begin" | "after_call" | null): boolean { if (crashAt === "after_begin") { this.up = false; return true; } return false; }
  private completeSuccess(run: RunState, step: StepState): void { step.status = "succeeded"; if (run.steps.every(s => s.status === "succeeded")) run.status = "succeeded"; }
  private transient(run: RunState, step: StepState): void {
    if (step.attempts < this.workflow.maxAttempts) { const deadline = this.now + this.workflow.retryDelay; require(deadline <= TIME_LIMIT, "TIME_OVERFLOW"); step.status = "pending"; step.ready_at = deadline; }
    else { step.status = "failed"; run.status = "failed"; for (const other of run.steps) if (other.status === "pending") other.status = "blocked"; }
  }
  tick(crashAt: "after_begin" | "after_call" | null): void {
    const a = this.selectAction(); if (!a) return; const { run, step, definition } = a;
    if (run.status === "cancelling") {
      if (this.checkpoint(crashAt)) return;
      const outcome = this.service.lookup([run.id, step.id], step.attempts);
      if (crashAt === "after_call") { this.up = false; return; }
      step.status = outcome === "found" ? "succeeded" : "cancelled"; run.status = "cancelled"; return;
    }
    if (step.status === "pending") { step.status = "running"; step.attempts++; }
    if (this.checkpoint(crashAt)) return;
    const outcome = this.service.execute([run.id, step.id], definition.amount, step.attempts, definition.failures);
    if (crashAt === "after_call") { this.up = false; return; }
    if (outcome === "transient") this.transient(run, step); else this.completeSuccess(run, step);
  }
  private cancel(id: string): void {
    const run = this.runs.find(r => r.id === id); require(run, "UNKNOWN_RUN");
    if (run.status === "succeeded" || run.status === "failed" || run.status === "cancelled") return;
    run.cancel_requested = true; for (const step of run.steps) if (step.status === "pending") step.status = "cancelled";
    run.status = run.steps.some(s => s.status === "running") ? "cancelling" : "cancelled";
  }
  apply(command: Command): void {
    if (command.op === "observe") { this.observations.push(this.snapshot()); return; }
    if (command.op === "advance") { require(this.now + command.by <= TIME_LIMIT, "TIME_OVERFLOW"); this.now += command.by; return; }
    if (command.op === "restart") { require(!this.up, "PROCESS_UP"); this.up = true; return; }
    require(this.up, "PROCESS_DOWN");
    if (command.op === "crash") { this.up = false; return; }
    if (command.op === "start") { require(!this.runs.some(r => r.id === command.run), "DUPLICATE_RUN"); this.runs.push({ id: command.run, status: "active", cancel_requested: false, steps: this.workflow.steps.map(s => ({ id: s.id, status: "pending", attempts: 0, ready_at: this.now })) }); return; }
    if (command.op === "cancel") { this.cancel(command.run); return; }
    if (command.op === "tick") this.tick(command.crashAt);
  }
  snapshot(): Snapshot { return { now: this.now, up: this.up, runs: this.runs.map(r => ({ ...r, steps: r.steps.map(s => ({ ...s })) })) }; }
  run(): SimulationResult { for (const command of this.workflow.commands) this.apply(command); return { observations: this.observations, final: { ...this.snapshot(), calls: this.service.calls, effects: this.service.effects } }; }
}

File: eval-reviews/2026-09-12-chained/workflow-recovery/python/terra/r1/starter/workflow.py
"""Deterministic durable workflow simulators (original and leased-worker modes)."""
from dataclasses import dataclass, field
import re
from typing import Any

TIME_LIMIT = 2_147_483_647


class DomainError(Exception):
    def __init__(self, code: str): self.code = code; super().__init__(code)


def require(ok: bool, code: str = "INVALID_INPUT") -> None:
    if not ok: raise DomainError(code)


def fields(v: Any, required: set[str], optional: set[str] = frozenset()) -> dict:
    require(type(v) is dict and required <= v.keys() <= required | optional)
    return v


def integer(v: Any, low: int, high: int) -> int:
    require(type(v) is int and low <= v <= high)
    return v


def ident(v: Any) -> str:
    require(type(v) is str and re.fullmatch(r"[A-Za-z0-9_-]{1,64}", v) is not None)
    return v


@dataclass(frozen=True)
class StepDef:
    id: str; needs: tuple[str, ...]; amount: int; failures: int = 0


def parse_step(v: Any) -> StepDef:
    o = fields(v, {"id", "needs", "amount"}, {"failures"})
    require(type(o["needs"]) is list)
    needs = tuple(ident(x) for x in o["needs"])
    require(len(needs) == len(set(needs)))
    return StepDef(ident(o["id"]), needs, integer(o["amount"], 1, 1_000_000),
                   integer(o.get("failures", 0), 0, 100))


def validate_graph(steps: tuple[StepDef, ...]) -> None:
    names = {x.id for x in steps}
    require(len(names) == len(steps), "DUPLICATE_STEP")
    require(all(n in names for x in steps for n in x.needs), "UNKNOWN_DEPENDENCY")
    seen: set[str] = set(); todo = list(steps)
    while todo:
        ready = [x for x in todo if set(x.needs) <= seen]
        require(bool(ready), "DEPENDENCY_CYCLE")
        seen.update(x.id for x in ready)
        todo = [x for x in todo if x.id not in seen]


@dataclass
class Step:
    id: str; status: str = "pending"; attempts: int = 0; ready_at: int = 0


@dataclass
class Run:
    id: str; steps: list[Step]; status: str = "active"; cancel_requested: bool = False


@dataclass
class Service:
    effects: list[tuple[tuple[str, str], int]] = field(default_factory=list)
    failures: dict[tuple[str, str], int] = field(default_factory=dict)

    def execute(self, key, amount, failures):
        if any(k == key for k, _ in self.effects): return "replayed"
        if self.failures.get(key, 0) < failures:
            self.failures[key] = self.failures.get(key, 0) + 1
            return "transient"
        self.effects.append((key, amount)); return "applied"

    def lookup(self, key):
        return "found" if any(k == key for k, _ in self.effects) else "missing"


# Checkpoint one (kept as the no-workers protocol).
@dataclass(frozen=True)
class OldCmd:
    op: str; run: str | None = None; by: int = 0; crash_at: str | None = None


def parse_old_cmd(v: Any) -> OldCmd:
    require(type(v) is dict and type(v.get("op")) is str); op = v["op"]
    if op in ("start", "cancel"):
        fields(v, {"op", "run"}); return OldCmd(op, ident(v["run"]))
    if op == "advance": fields(v, {"op", "by"}); return OldCmd(op, by=integer(v["by"], 0, TIME_LIMIT))
    if op == "tick":
        fields(v, {"op"}, {"crash_at"}); c = v.get("crash_at")
        if "crash_at" in v:
            require(type(c) is str and c in ("after_begin", "after_call"))
        return OldCmd(op, crash_at=c)
    require(op in ("observe", "crash", "restart")); fields(v, {"op"}); return OldCmd(op)


class OldSimulator:
    def __init__(self, steps, cmds, max_attempts, retry_delay):
        self.steps_def, self.cmds, self.max_attempts, self.retry_delay = steps, cmds, max_attempts, retry_delay
        self.now = 0; self.up = True; self.runs: list[Run] = []; self.service = Service(); self.observations = []; self.calls = []
    def select(self):
        for r in self.runs:
            for s, d in zip(r.steps, self.steps_def):
                if s.status == "running": return r, s, d
        for r in self.runs:
            if r.status == "active":
                done = {s.id for s in r.steps if s.status == "succeeded"}
                for s, d in zip(r.steps, self.steps_def):
                    if s.status == "pending" and s.ready_at <= self.now and all(x in done for x in d.needs): return r, s, d
        return None
    def finish_success(self, r, s):
        s.status = "succeeded"
        if all(x.status == "succeeded" for x in r.steps): r.status = "succeeded"
    def transient(self, r, s):
        if s.attempts < self.max_attempts:
            require(self.now + self.retry_delay <= TIME_LIMIT, "TIME_OVERFLOW")
            s.status = "pending"; s.ready_at = self.now + self.retry_delay
        else:
            s.status = "failed"; r.status = "failed"
            for x in r.steps:
                if x.status == "pending": x.status = "blocked"
    def tick(self, crash_at):
        a = self.select()
        if not a: return
        r, s, d = a
        if s.status == "pending": s.status = "running"; s.attempts += 1
        if crash_at == "after_begin": self.up = False; return
        key = (r.id, s.id)
        if r.status == "cancelling":
            out = self.service.lookup(key); self.calls.append(("lookup", key, s.attempts, out))
            if crash_at == "after_call": self.up = False; return
            s.status = "succeeded" if out == "found" else "cancelled"; r.status = "cancelled"; return
        out = self.service.execute(key, d.amount, d.failures); self.calls.append(("execute", key, s.attempts, out))
        if crash_at == "after_call": self.up = False; return
        self.finish_success(r, s) if out != "transient" else self.transient(r, s)
    def cancel(self, rid):
        r = next((x for x in self.runs if x.id == rid), None); require(r is not None, "UNKNOWN_RUN")
        if r.status in ("succeeded", "failed", "cancelled"): return
        r.cancel_requested = True
        for s in r.steps:
            if s.status == "pending": s.status = "cancelled"
        r.status = "cancelling" if any(s.status == "running" for s in r.steps) else "cancelled"
    def snapshot(self):
        return {"now": self.now, "up": self.up, "runs": [{"id": r.id, "status": r.status, "cancel_requested": r.cancel_requested, "steps": [{"id": s.id, "status": s.status, "attempts": s.attempts, "ready_at": s.ready_at} for s in r.steps]} for r in self.runs]}
    def apply(self, c):
        if c.op == "observe": self.observations.append(self.snapshot()); return
        if c.op == "advance": require(self.now + c.by <= TIME_LIMIT, "TIME_OVERFLOW"); self.now += c.by; return
        if c.op == "restart": require(not self.up, "PROCESS_UP"); self.up = True; return
        require(self.up, "PROCESS_DOWN")
        if c.op == "crash": self.up = False
        elif c.op == "start":
            require(all(r.id != c.run for r in self.runs), "DUPLICATE_RUN")
            self.runs.append(Run(c.run, [Step(d.id, ready_at=self.now) for d in self.steps_def]))
        elif c.op == "cancel": self.cancel(c.run)
        else: self.tick(c.crash_at)
    def run(self):
        for c in self.cmds: self.apply(c)
        final = self.snapshot(); final["calls"] = [{"kind": k, "key": list(key), "attempt": a, "outcome": o} for k,key,a,o in self.calls]
        final["effects"] = [{"key": list(k), "amount": a} for k,a in self.service.effects]
        return {"observations": self.observations, "final": final}


@dataclass
class Lease: worker: str; ticket: int; expires: int
@dataclass
class LStep(Step): lease: Lease | None = None
@dataclass
class Ticket: id: int; worker: str; run: Run; step: LStep; definition: StepDef; kind: str; response: str | None = None; delivered: bool = False


def parse_leased_cmd(v):
    require(type(v) is dict and type(v.get("op")) is str); op = v["op"]
    if op in ("start", "cancel"): fields(v, {"op", "run"}); return (op, ident(v["run"]))
    if op == "advance": fields(v, {"op", "by"}); return (op, integer(v["by"], 0, TIME_LIMIT))
    if op in ("observe",): fields(v, {"op"}); return (op,)
    if op in ("crash", "restart", "claim"): fields(v, {"op", "worker"}); return (op, ident(v["worker"]))
    if op in ("renew", "call"):
        fields(v, {"op", "worker", "ticket"}); return (op, ident(v["worker"]), integer(v["ticket"], 1, TIME_LIMIT))
    if op == "deliver": fields(v, {"op", "ticket"}); return (op, integer(v["ticket"], 1, TIME_LIMIT))
    raise DomainError("INVALID_INPUT")


class LeasedSimulator:
    def __init__(self, steps, cmds, workers, max_attempts, retry_delay, duration):
        self.defs, self.cmds, self.max_attempts, self.retry_delay, self.duration = steps, cmds, max_attempts, retry_delay, duration
        self.workers = {w: True for w in workers}; self.worker_order = workers; self.now = 0; self.runs = []; self.service = Service(); self.tickets = {}; self.next_ticket = 1; self.calls = []
    def worker(self, w, up=True):
        require(w in self.workers, "UNKNOWN_WORKER")
        if up: require(self.workers[w], "WORKER_DOWN")
    def ticket(self, n): require(n in self.tickets, "UNKNOWN_TICKET"); return self.tickets[n]
    def owned(self, w, n):
        self.worker(w); t = self.ticket(n); require(t.worker == w, "WRONG_WORKER"); return t
    def live(self, lease): return lease is not None and self.now < lease.expires
    def select(self):
        for r in self.runs:
            for s,d in zip(r.steps,self.defs):
                if s.status == "running" and not self.live(s.lease): return r,s,d
        for r in self.runs:
            if r.status == "active":
                done={s.id for s in r.steps if s.status=="succeeded"}
                for s,d in zip(r.steps,self.defs):
                    if s.status=="pending" and s.ready_at<=self.now and all(x in done for x in d.needs): return r,s,d
        return None
    def claim(self,w):
        self.worker(w)
        require(not any(s.status=="running" and self.live(s.lease) and s.lease.worker==w for r in self.runs for s in r.steps), "WORKER_BUSY")
        item=self.select()
        if not item:return {"ticket":None}
        r,s,d=item; require(self.now+self.duration<=TIME_LIMIT,"TIME_OVERFLOW")
        if s.status=="pending": s.status="running"; s.attempts+=1
        n=self.next_ticket; self.next_ticket+=1; s.lease=Lease(w,n,self.now+self.duration)
        kind="execute" if r.status=="active" else "lookup"
        self.tickets[n]=Ticket(n,w,r,s,d,kind); return {"ticket":n}
    def call(self,w,n):
        t=self.owned(w,n)
        if t.step.lease is None or t.step.lease.ticket != n or not self.live(t.step.lease): return {"outcome":"stale"}
        if t.response is None:
            key=(t.run.id,t.step.id)
            t.response=self.service.execute(key,t.definition.amount,t.definition.failures) if t.kind=="execute" else self.service.lookup(key)
            self.calls.append({"worker":w,"ticket":n,"kind":t.kind,"key":list(key),"attempt":t.step.attempts,"outcome":t.response})
        return {"kind":t.kind,"outcome":t.response}
    def settle_terminal(self,r):
        if r.status=="cancelling" and not any(s.status=="running" for s in r.steps): r.status="cancelled"
        if r.status=="failing" and not any(s.status=="running" for s in r.steps): r.status="failed"
    def deliver(self,n):
        t=self.ticket(n); s=t.step; lease=s.lease
        if t.response is None or t.delivered or lease is None or lease.ticket!=n or not self.live(lease) or not self.workers[t.worker]: return {"committed":False}
        t.delivered=True; s.lease=None; r=t.run; out=t.response
        if t.kind=="lookup":
            if r.status=="cancelling": s.status="succeeded" if out=="found" else "cancelled"
            else: s.status="succeeded" if out=="found" else "blocked"
            self.settle_terminal(r); return {"committed":True}
        if out in ("applied","replayed"):
            s.status="succeeded"
            if r.status=="active" and all(x.status=="succeeded" for x in r.steps): r.status="succeeded"
        elif s.attempts < self.max_attempts:
            require(self.now+self.retry_delay<=TIME_LIMIT,"TIME_OVERFLOW"); s.status="pending"; s.ready_at=self.now+self.retry_delay
        else:
            s.status="failed"
            for x in r.steps:
                if x.status=="pending": x.status="blocked"
            running=[x for x in r.steps if x.status=="running"]
            if running:
                r.status="failing"
                for x in running: x.lease.expires=self.now
            else:r.status="failed"
        return {"committed":True}
    def cancel(self,rid):
        r=next((x for x in self.runs if x.id==rid),None); require(r is not None,"UNKNOWN_RUN")
        if r.status in ("succeeded","failed","cancelled","failing","cancelling"): return
        r.cancel_requested=True
        for s in r.steps:
            if s.status=="pending":s.status="cancelled"
        running=[s for s in r.steps if s.status=="running"]
        if running:
            r.status="cancelling"
            for s in running:s.lease.expires=self.now
        else:r.status="cancelled"
    def snapshot(self):
        return {"now":self.now,"workers":[{"id":w,"up":self.workers[w]} for w in self.worker_order],"runs":[{"id":r.id,"status":r.status,"cancel_requested":r.cancel_requested,"steps":[{"id":s.id,"status":s.status,"attempts":s.attempts,"ready_at":s.ready_at,"lease":None if s.lease is None else {"worker":s.lease.worker,"ticket":s.lease.ticket,"expires":s.lease.expires}} for s in r.steps]} for r in self.runs]}
    def apply(self,c):
        op=c[0]
        if op=="observe":return self.snapshot()
        if op=="advance":require(self.now+c[1]<=TIME_LIMIT,"TIME_OVERFLOW");self.now+=c[1];return None
        if op=="start":
            require(all(r.id!=c[1] for r in self.runs),"DUPLICATE_RUN");self.runs.append(Run(c[1],[LStep(d.id,ready_at=self.now) for d in self.defs]));return None
        if op=="cancel":self.cancel(c[1]);return None
        if op=="crash":self.worker(c[1]);self.workers[c[1]]=False;return None
        if op=="restart":
            require(c[1] in self.workers,"UNKNOWN_WORKER");require(not self.workers[c[1]],"WORKER_UP");self.workers[c[1]]=True;return None
        if op=="claim":return self.claim(c[1])
        if op=="renew":
            t=self.owned(c[1],c[2]); s=t.step
            if s.lease is not None and s.lease.ticket==t.id and self.live(s.lease):
                require(self.now+self.duration<=TIME_LIMIT,"TIME_OVERFLOW");s.lease.expires=self.now+self.duration;return {"renewed":True}
            return {"renewed":False}
        if op=="call":return self.call(c[1],c[2])
        return self.deliver(c[1])
    def run(self):
        values=[self.apply(c) for c in self.cmds]; final=self.snapshot();final["calls"]=self.calls;final["effects"]=[{"key":list(k),"amount":a} for k,a in self.service.effects]
        return {"results":values,"final":final}


def parse_workflow(raw: Any):
    require(type(raw) is dict)
    leased="workers" in raw
    allowed={"steps","commands","max_attempts","retry_delay"} | ({"workers","lease_duration"} if leased else set())
    fields(raw,{"steps","commands"} | ({"workers"} if leased else set()),allowed-({"steps","commands"} | ({"workers"} if leased else set())))
    require(type(raw["steps"]) is list and bool(raw["steps"]) and type(raw["commands"]) is list)
    steps=tuple(parse_step(x) for x in raw["steps"])
    maxa=integer(raw.get("max_attempts",3),1,10); delay=integer(raw.get("retry_delay",2),1,1_000_000)
    if not leased:
        cmds=tuple(parse_old_cmd(x) for x in raw["commands"]); validate_graph(steps); return OldSimulator(steps,cmds,maxa,delay)
    require(len(raw["commands"])<=2000 and type(raw["workers"]) is list and 1<=len(raw["workers"])<=100)
    workers=tuple(ident(x) for x in raw["workers"]);require(len(workers)==len(set(workers)))
    duration=integer(raw.get("lease_duration",5),1,1_000_000)
    cmds=tuple(parse_leased_cmd(x) for x in raw["commands"]); validate_graph(steps)
    return LeasedSimulator(steps,cmds,workers,maxa,delay,duration)


# Main imports these names; parse_workflow now returns the appropriate simulator.
Simulator = object
